source/common/stats/custom_namespace.cc

```diff
@@ -0,0 +1,25 @@
+#include "source/common/stats/custom_namespace.h"
+
+#include "envoy/registry/registry.h"
+
+namespace Envoy {
+namespace Stats {
+
+void CustomStatNamespaceFactoryImpl::registerStatNamespace(const absl::string_view& name) {
+  namespaces_.insert(std::string(name));
+};
+
+std::string
+CustomStatNamespaceFactoryImpl::trySanitizeStatName(const absl::string_view& stat_name) const {
+  const auto pos = stat_name.find_first_of('.');
+  if (namespaces_.find(stat_name.substr(0, pos)) != namespaces_.end()) {
+    // Trim the custom namespace.
+    return std::string(stat_name.substr(pos + 1));
+  }
+  return "";
+};
+
+REGISTER_FACTORY(CustomStatNamespaceFactoryImpl, CustomStatNamespaceFactory);
```

Comment:
I think we can probably avoid a factory here and just use a thread safe singleton or something like that?

Reply:
If you want to use a singleton that's OK as long as it's resets between every test method, right?

What I was thinking was that no new static mutable state was needed (my usual bias) but that you could hang this new object off an existing factory that got passed into an extension initializer, but I don't know the details of what's easy to do.

Reply:
After I introduced the factory in the latest commit, I see that it is the same situation for both of them - changing the signature of functions so that this object is passed as an argument would be the only way to testability, rather than directory accessing to the global state either via Registry::FactoryRegistry or a static mutable state getter in the 
test target functions.

I will come back to the static mutable state of CustomStatNamespaces object again and let's see how the tests will look like.

Reply:
I hope we can get this done without any static mutable state or new global factories (as I said earlier). I realize it's less lines of code to do it that way.

Changing function signatures to pass the right state to the right places sounds good to me. It's more editing of code but easier to reason about. Happy to defer to Matt on this decision of course. If we decide do employ static mutable state then be sure to follow patterns in Envoy to make tests hermetic.

Reply:
Sure if we can avoid statics please do that. I was mostly saying that I would prefer to not use a "factory" as that shows up in various logging, etc. For this it should just be a "basic static" of whatever type we prefer.